Add support for optional AuthenticatorSelection

samalone opened the pull request:
First, I want to be up-front that this PR was created with AI assistance. My understanding of WebAuthn is limited.

The purpose of this PR is to add optional `AuthenticatorSelection` support to `PublicKeyCredentialCreationOptions`. Without authenticator selection, Chrome on Android 16 devices always creates device-locked passkeys that are not discoverable. By specifying that the server requires a client-side-resident credential, you can get Android to create the passkey in Google Password Manager, which makes it discoverable.

Using this, I was able to get Android to support a passkey workflow that matches Safari.

samalone committed:
Add authenticator selection criteria to registration options

- Introduced `AuthenticatorSelection` struct to define Relying Party's requirements for authenticator attributes.
- Updated `PublicKeyCredentialCreationOptions` to include `authenticatorSelection` parameter.
- Enhanced `WebAuthnManager`'s `beginRegistration` method to accept `authenticatorSelection` as an argument.
- Added `ResidentKeyRequirement` struct to specify requirements for client-side-resident public key credentials.

samalone committed:
Refactor authenticator selection handling in registration process

- Enhanced `WebAuthnManager` to better manage authenticator selection during registration.
- Updated related structs and methods to improve clarity and maintainability.
- Added tests to ensure correct functionality of the new authenticator selection features.

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/AuthenticatorSelection.swift:
```diff
@@ -0,0 +1,39 @@
+/// A dictionary describing the Relying Party's requirements regarding authenticator attributes.
+///
+/// - SeeAlso: https://www.w3.org/TR/webauthn-2/#dictdef-authenticatorselectioncriteria
+public struct AuthenticatorSelection: Codable, Sendable {
```
Could you conform to `Hashable` while you are at it?

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/AuthenticatorSelection.swift:
```diff
@@ -0,0 +1,39 @@
+    public var authenticatorAttachment: AuthenticatorAttachment?
+    
+    /// Describes the Relying Party's requirements regarding whether the authenticator should create a client-side-resident public key credential source.
+    public var residentKey: ResidentKeyRequirement?
```
Can you add an explicit encoder implementation (as an extension) that does the following?
> **requireResidentKey, of type [boolean](https://heycam.github.io/webidl/#idl-boolean), defaulting to false**
> This member is retained for backwards compatibility with WebAuthn Level 1 and, for historical reasons, its naming retains the deprecated “resident” terminology for [discoverable credentials](https://www.w3.org/TR/webauthn-2/#discoverable-credential). [Relying Parties](https://www.w3.org/TR/webauthn-2/#relying-party) SHOULD set it to true if, and only if, [residentKey](https://www.w3.org/TR/webauthn-2/#dom-authenticatorselectioncriteria-residentkey) is set to [required](https://www.w3.org/TR/webauthn-2/#dom-residentkeyrequirement-required).

https://www.w3.org/TR/webauthn-2/#dom-authenticatorselectioncriteria-requireresidentkey

dimitribouniol replied:
```suggestion
    /// Describes the Relying Party's requirements regarding whether the authenticator should create a client-side-resident public key credential source.
    /// - SeeAlso: [WebAuthn Level 3 Working Draft §5.4.4. Authenticator Selection Criteria](https://www.w3.org/TR/webauthn-3/#dom-authenticatorselectioncriteria-residentkey)
    public var residentKey: ResidentKeyRequirement?
```

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/AuthenticatorSelection.swift:
```diff
@@ -0,0 +1,39 @@
+
+/// A dictionary describing the Relying Party's requirements regarding authenticator attributes.
+///
+/// - SeeAlso: https://www.w3.org/TR/webauthn-2/#dictdef-authenticatorselectioncriteria
```
For the SeeAlso refs, let's go with the more detailed link style:
```suggestion
/// - SeeAlso: [WebAuthn Level 3 Working Draft §5.4.4. Authenticator Selection Criteria](https://www.w3.org/TR/webauthn-3/#dictionary-authenticatorSelection)
```

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/AuthenticatorSelection.swift:
```diff
@@ -0,0 +1,39 @@
+/// A dictionary describing the Relying Party's requirements regarding authenticator attributes.
+///
+/// - SeeAlso: https://www.w3.org/TR/webauthn-2/#dictdef-authenticatorselectioncriteria
+public struct AuthenticatorSelection: Codable, Sendable {
+    /// If present, indicates the Relying Party's preference for authenticator attachment.
+    public var authenticatorAttachment: AuthenticatorAttachment?
```
```suggestion
    /// If present, indicates the Relying Party's preference for authenticator attachment.
    /// - SeeAlso: [WebAuthn Level 3 Working Draft §5.4.4. Authenticator Selection Criteria](https://www.w3.org/TR/webauthn-3/#dom-authenticatorselectioncriteria-authenticatorattachment)
    public var authenticatorAttachment: AuthenticatorAttachment?
```

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/AuthenticatorSelection.swift:
```diff
@@ -0,0 +1,39 @@
+    
+    /// Describes the Relying Party's requirements regarding whether the authenticator should create a client-side-resident public key credential source.
+    public var residentKey: ResidentKeyRequirement?
+    
+    /// Describes the Relying Party's requirements regarding user verification.
+    public var userVerification: UserVerificationRequirement?
```
```suggestion
    /// Describes the Relying Party's requirements regarding user verification.
    /// - SeeAlso: [WebAuthn Level 3 Working Draft §5.4.4. Authenticator Selection Criteria](https://www.w3.org/TR/webauthn-3/#dom-authenticatorselectioncriteria-userverification)
    public var userVerification: UserVerificationRequirement?
```

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/ResidentKeyRequirement.swift:
```diff
@@ -0,0 +1,36 @@
+
+/// The Relying Party's requirements regarding whether the authenticator should create a client-side-resident public key credential source.
+///
+/// - SeeAlso: https://www.w3.org/TR/webauthn-2/#enumdef-residentkeyrequirement
```
```suggestion
/// - SeeAlso: [WebAuthn Level 3 Working Draft §5.4.6. Resident Key Requirement Enumeration](https://www.w3.org/TR/webauthn-3/#enum-residentKeyRequirement)
```

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/ResidentKeyRequirement.swift:
```diff
@@ -0,0 +1,36 @@
+    
+    public init(_ rawValue: String) {
+        self.rawValue = rawValue
+    }
+    
+    /// This value indicates the Relying Party requires a client-side-resident credential (i.e., a discoverable credential).
+    /// If the authenticator cannot create a client-side-resident credential, it will return an error.
+    public static let required: Self = "required"
```
```suggestion
    /// This value indicates the Relying Party requires a client-side-resident credential (i.e., a discoverable credential).
    ///
    /// If the authenticator cannot create a client-side-resident credential, it will return an error.
    /// - SeeAlso: [WebAuthn Level 3 Working Draft §5.4.6. Resident Key Requirement Enumeration](https://www.w3.org/TR/webauthn-3/#dom-residentkeyrequirement-required)
    public static let required: Self = "required"
```

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/ResidentKeyRequirement.swift:
```diff
@@ -0,0 +1,36 @@
+    public static let required: Self = "required"
+    
+    /// This value indicates the Relying Party strongly prefers a client-side-resident credential, but will accept a server-side credential.
+    public static let preferred: Self = "preferred"
```
```suggestion
    /// This value indicates the Relying Party strongly prefers a client-side-resident credential, but will accept a server-side credential.
    /// - SeeAlso: [WebAuthn Level 3 Working Draft §5.4.6. Resident Key Requirement Enumeration](https://www.w3.org/TR/webauthn-3/#dom-residentkeyrequirement-preferred)
    public static let preferred: Self = "preferred"
```

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/ResidentKeyRequirement.swift:
```diff
@@ -0,0 +1,36 @@
+    
+    /// This value indicates the Relying Party strongly prefers a client-side-resident credential, but will accept a server-side credential.
+    public static let preferred: Self = "preferred"
+    
+    /// This value indicates the Relying Party strongly prefers a server-side credential, but will accept a client-side-resident credential.
+    public static let discouraged: Self = "discouraged"
```
```suggestion
    /// This value indicates the Relying Party strongly prefers a server-side credential, but will accept a client-side-resident credential.
    /// - SeeAlso: [WebAuthn Level 3 Working Draft §5.4.6. Resident Key Requirement Enumeration](https://www.w3.org/TR/webauthn-3/#dom-residentkeyrequirement-discouraged)
    public static let discouraged: Self = "discouraged"
```

dimitribouniol commented on Sources/WebAuthn/WebAuthnManager.swift:
```diff
@@ -57,12 +57,15 @@ public struct WebAuthnManager: Sendable {
     ///   - attestation: The Relying Party's preference regarding attestation. Defaults to `.none`.
     ///   - publicKeyCredentialParameters: A list of public key algorithms the Relying Party chooses to restrict
     ///     support to. Defaults to all supported algorithms.
+    ///   - authenticatorSelection: A dictionary describing the Relying Party's requirements regarding authenticator attributes.
```
```suggestion
    ///   - authenticatorSelection: The Relying Party's authenticator selection criteria that should be communicated to the client when choosing an authenticator to use.
```

dimitribouniol reviewed (changes requested):
Looks good for the most part! I wouldn't worry about the breaking change here, it would be limited to key path users, and the library is still in beta.

Other than that, please see my suggestions for the doc comments so we have something nicer to link folks out to, and add support for the `requireResidentKey` encoded property that is required by the spec so we don't forget to add it later.

Thanks!

samalone committed:
Enhance AuthenticatorSelection struct for improved decoding and encoding

- Updated `AuthenticatorSelection` to include custom encoding and decoding logic for `requireResidentKey`.
- Improved documentation with references to the WebAuthn Level 3 Working Draft.
- Added tests to validate the behavior of `requireResidentKey` during encoding and decoding processes.

dimitribouniol commented on Tests/WebAuthnTests/AuthenticatorSelectionTests.swift:
```diff
@@ -107,6 +107,48 @@ struct AuthenticatorSelectionTests {
         #expect(jsonString.contains("required"))
         #expect(jsonString.contains("preferred"))
+        #expect(jsonString.contains("requireResidentKey"))
+        #expect(jsonString.contains("true"))
```
Just noticed we are doing contains comparisons for the encoding — could we set `sortedKeys` and do a whole string comparison instead?

```suggestion
        let encoder = JSONEncoder()
        encoder.outputFormatting = .sortedKeys
        let json = try encoder.encode(selection)
        let jsonString = String(decoding: json, as: UTF8.self)
        #expect(jsonString == #"{...}"#)
```

Same for the other tests.

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/AuthenticatorSelection.swift:
```diff
@@ -37,3 +40,39 @@ public struct AuthenticatorSelection: Codable, Sendable {
     }
 }
 
+extension AuthenticatorSelection {
```
For consistency with other types, can we move the `Codable` conformance over to this line?
```suggestion
extension AuthenticatorSelection: Codable {
```

dimitribouniol reviewed (approved):
Just minor nits left, but otherwise looks good to me!

samalone committed:
Refactor AuthenticatorSelection struct to streamline Codable conformance

- Removed Codable conformance from the main struct and implemented it in an extension for better separation of concerns.
- Updated tests to ensure JSON encoding produces sorted keys and matches expected output format.

samalone committed:
Merge branch 'main' into authenticatorSelection-upstream

dimitribouniol commented:
Thanks!
